packages/core/src/app/block-production/merkle-tree.ts

```diff
@@ -127,11 +127,36 @@ export class SparseMerkleTreeImpl implements SparseMerkleTree {
   }
 
   public async update(leafKey: BigNumber, leafValue: Buffer): Promise<boolean> {
-    return this.treeMutex.runExclusive(async () => {
-      const nodesToUpdate: MerkleTreeNode[] = await this.getNodesInPath(leafKey)
-      if (!nodesToUpdate || nodesToUpdate.length !== this.height) {
+    let nodesToUpdate: MerkleTreeNode[] = await this.treeMutex.runExclusive(
+      () => {
+        return this.getNodesInPath(leafKey)
+      }
+    )
+
+    if (!nodesToUpdate) {
+      return false
+    } else if (nodesToUpdate.length !== this.height) {
+      if (
+        !(await this.verifyAndStorePartiallyEmptyPath(
+          leafKey,
+          nodesToUpdate.length
+        ))
+      ) {
         return false
       }
+      nodesToUpdate = await this.getNodesInPath(leafKey)
+    }
+
+    return this.treeMutex.runExclusive(async () => {
+      // Have to check to make sure nodesToUpdate didn't change while we didn't have the lock.
```

Comment:
I am probably missing some stuff but why not keep a lock the entire time that you're doing the update? Just to reduce a little complexity in this code. Is that possible?

Reply:
There's another call that locks, so this would deadlock if the lock is held. Other languages have the concept of re-entrant locks where if your thread is the one holding the lock, you can access anything protected by a lock, but unfortunately the mutex library doesn't do this, and my limited research shows that there's no easy swap-in replacement for it that does (since JS is single-threaded, and it may be hard to keep track of parallel contexts).

Reply:
That said, I agree with you and am not happy with how ugly this looks.

Reply:
Ok that's fair. Then well it sounds like you did the due diligence here & so imo let's just roll with the ugly code & clean it up later as at least this technical debt is isolated.